DEVPROD-4593 Default to enabling optimisations and using release vcpkg libs

thessem opened the pull request:
**Ticket**: [DEVPROD-4593](https://jira.mongodb.org/browse/DEVPROD-4593): _Fix Genny compiling without optimization and linking against libraries with debug symbols_

Genny's compilation does not set a default `CMAKE_BUILD_TYPE`, meaning it doesn't build with any optimisations set, nor does it build with debug symbols. Regrettably, when `vcpkg` is imported, the lack of a default causes CMake to link against the `gennytoolchain` debug libraries.

The requirement to set `CMAKE_BUILD_TYPE=Release` for build optimisations doesn't seem to be widely known among users of Genny, and linking the toolchain against debug libraries is inappropriate as a default behavior.

#### What's changed

- The generated `vcpkg` toolchain is imported into Genny's CMake system. This allows it to recognize any variables set in the CMake process along with any variables set in the CLI arguments.
- If not already set, the `CMAKE_BUILD_TYPE` defaults to `Release`.

#### Testing

Compile time with this PR was measured at ~2 min 40s ([here](https://evergreen.mongodb.com/task_log_raw/sys_perf_perf_3_node_replSet.arm.aws.2023_11_mixed_workloads_genny_rate_limited_high_value_patch_48a18fad18098b7ef67e65f763382ddc75320e47_65cda4883e8e86c4da93818b_24_02_15_05_46_24/0?type=T)) and without at ~1 min 40s ([here](https://spruce.mongodb.com/task/sys_perf_perf_3_node_replSet.arm.aws.2023_11_mixed_workloads_genny_5757f2e4ce595e3de14e6cc91d972ec9d1678f40_24_02_15_11_59_31/logs?execution=0&sortBy=STATUS&sortDir=ASC)). This was on a `r6g.2xlarge` with 5000 IOPS. This is 1 minute of extra overhead compiling.

There is a Performance Analyser test against high-value workloads available [here](https://performance-analyzer.server-tig.prod.corp.mongodb.com/perf-analyzer-viz?comparison_id=665c50c2-5c89-4181-984d-572eae14a543). I observed a change greater than 0.5 standard deviations in several metrics.

| **variant**                                     | **task**                                              | **test**                                    | **args**                               | **measurement**                |     **z_score** |
|:--------------------------------------------|:--------------------------------------------------|:----------------------------------------|:-----------------------------------|:---------------------------|------------:|
| perf-3-node-replSet.arm.aws-11         | expressive_queries_high_value                     | RunQueries.QueryWithProjectionWithIndex | {}                                 | Latency95thPercentile      |    -3.00382 |
| perf-3-node-replSet.arm.aws-11         | expressive_queries_high_value                     | RunQueries.QueryWithProjectionWithIndex | {}                                 | OperationThroughput        |     12.6143 |
| perf-3-node-replSet.arm.aws-11         | large_indexed_ins_high_value                      | FindLargeIn.find                        | {}                                 | OperationThroughput        |     1.09022 |
| perf-3-node-replSet.arm.aws-11         | load_test_high_value                              | InitialNoIndexes.TotalBulkInsert        | {}                                 | Latency95thPercentile      |    -14.4569 |
| perf-3-node-replSet.arm.aws-11         | load_test_high_value                              | InitialNoIndexes.TotalBulkInsert        | {}                                 | OperationThroughput        |     16.5204 |
| perf-3-node-replSet.arm.aws-11         | majority_reads10_k_threads_high_value             | WritingActor_90_10.WriteDocs            | {}                                 | Latency95thPercentile      |   -0.606211 |
| perf-3-node-replSet.arm.aws-11         | majority_reads10_k_threads_high_value             | ReadingActor_90_10.ReadDocs             | {}                                 | Latency95thPercentile      |    0.508891 |
| perf-3-node-replSet.arm.aws-11         | majority_reads10_k_threads_high_value             | WritingActor_90_10.WriteDocs            | {}                                 | OperationThroughput        |    0.743113 |
| perf-shard-lite-fle.arm.aws-11         | medical_workload_diagnosis_50_50_high_value       | InsertActor.FSM.query                   | {}                                 | OperationThroughput        |     29.8414 |
| perf-3-node-replSet-intel.intel.aws-11 | mixed_workloads_genny_rate_limited_high_value     | mixed_workloads_genny_rate_limited      | {"genny_phase_id": 2, "mongod": 2} | system cpu user (%) - mean |   -0.601603 |
| perf-3-node-replSet-intel.intel.aws-11 | mixed_workloads_genny_rate_limited_high_value     | mixed_workloads_genny_rate_limited      | {"genny_phase_id": 2, "mongod": 1} | system cpu user (%) - mean |   -0.683848 |
| perf-3-node-replSet-intel.intel.aws-11 | mixed_workloads_genny_rate_limited_high_value     | Find_250.findOne                        | {}                                 | Latency50thPercentile      |    -2.17444 |
| perf-3-node-replSet.arm.aws-11         | mixed_workloads_genny_rate_limited_high_value     | mixed_workloads_genny_rate_limited      | {"genny_phase_id": 2, "mongod": 2} | system cpu user (%) - mean |    -1.24776 |
| perf-3-node-replSet.arm.aws-11         | mixed_workloads_genny_rate_limited_high_value     | mixed_workloads_genny_rate_limited      | {"genny_phase_id": 2, "mongod": 1} | system cpu user (%) - mean |    -1.10271 |
| perf-3-node-replSet.arm.aws-11         | mixed_workloads_genny_rate_limited_high_value     | mixed_workloads_genny_rate_limited      | {"genny_phase_id": 2, "mongod": 0} | system cpu user (%) - mean |   -0.647131 |
| perf-3-node-replSet.arm.aws-11         | mixed_workloads_genny_rate_limited_high_value     | Find_250.findOne                        | {}                                 | Latency95thPercentile      |   -0.897709 |
| perf-3-node-replSet.arm.aws-11         | mixed_workloads_genny_rate_limited_high_value     | Find_250.findOne                        | {}                                 | Latency50thPercentile      |    -2.40254 |
| perf-3-node-replSet.arm.aws-11         | mixed_workloads_genny_rate_limited_high_value     | mixed_workloads_genny_rate_limited      | {"genny_phase_id": 0, "mongod": 2} | system cpu user (%) - mean |    0.668843 |
| perf-3-node-replSet.arm.aws-11         | mixed_workloads_genny_rate_limited_high_value     | Update_250.updateOne                    | {}                                 | Latency95thPercentile      |   -0.733831 |
| perf-3-node-replSet.arm.aws-11         | mixed_workloads_genny_rate_limited_high_value     | Update_250.updateOne                    | {}                                 | Latency50thPercentile      |   -0.804057 |
| perf-3-node-replSet.arm.aws-11         | time_series_sort_high_value                       | Queries.SortQueryBoundedSort            | {}                                 | Latency50thPercentile      | -0.00468112 |
| perf-3-node-replSet.arm.aws-11         | time_series_sort_high_value                       | Queries.SortQueryBoundedSort            | {}                                 | OperationThroughput        |    0.090991 |
| perf-shard-lite-fle.arm.aws-11         | ycsb_like_queryable_encrypt1_cfdefault_high_value | ShardCollection.EnableSharding          | {}                                 | Latency95thPercentile      |   -0.591381 |
| perf-shard-lite-fle.arm.aws-11         | ycsb_like_queryable_encrypt1_cfdefault_high_value | YCSBLikeread50update.writes          | {}                                 | Latency95thPercentile      |    0.454819 |
| perf-shard-lite-fle.arm.aws-11         | ycsb_like_queryable_encrypt1_cfdefault_high_value | YCSBLikeread50update.writes          | {}                                 | OperationThroughput        |   -0.095896 |
| perf-shard-lite-fle.arm.aws-11         | ycsb_like_queryable_encrypt1_cfdefault_high_value | ShardCollection.EnableSharding          | {}                                 | OperationThroughput        |    0.477017 |

thessem commented:
I have requested a review from @mongodb/devprod-performance-infrastructure for the code, I have requested a review from @mongodb/performance for both the performance changes and the 1 minute increase in test setup time.

thessem commented:
@jimoleary Would you be able to offer any insight as to why [t_cmake_test](https://spruce.mongodb.com/task/genny_amazon2_arm64_t_cmake_test_patch_d103863204986e8caf63ef7d112d22beef7eda75_65cec0a4c9ec4424a0fdf7c3_24_02_16_01_55_49?execution=1) has started consistently failing on [Amazon Linux 2 arm64](https://spruce.mongodb.com/variant-history/genny/amazon2_arm64), but not [Amazon Linux 2 x64](https://spruce.mongodb.com/variant-history/genny/amazon2) or either macOS? Error is 

```
 [2024/02/16 13:15:31.036] 110/112 Test  #92: gennylib_test:single-threaded range-based for loops no phases blocking ...............................   Passed    0.01 sec
 [2024/02/16 13:15:31.796] 111/112 Test  #56: gennylib_test:PoolFactory behavior ...................................................................   Passed    2.02 sec
 [2024/02/16 13:15:32.140] 112/112 Test #119: value_generators_test:YAML Tests .....................................................................***Failed    2.31 sec
 [2024/02/16 13:15:32.140] [2024-02-16 02:15:31.186704] [0x0000ffff9d65edb0] [warning]  RandomDateGenerator ^RandomDate: {min: 2020-01-01, max: 2020-01-01}, max (1577836800000) must be greater than min (1577836800000)
 [2024/02/16 13:15:32.140] [2024-02-16 02:15:31.198911] [0x0000ffff9d65edb0] [warning]  RandomDateGenerator ^RandomDate: {min: 2021-01-01, max: 2020-01-01}, max (1577836800000) must be greater than min (1609459200000)
 [2024/02/16 13:15:32.140] [2024-02-16 02:15:31.852786] [0x0000ffff9d65edb0] [warning]  UuidGenerator '' is not a valid format. The format must match: '^([0-9a-f]{8}\-[0-9a-f]{4}\-[0-9a-f]{4}\-[0-9a-f]{4}\-[0-9a-f]{12})$'
 [2024/02/16 13:15:32.140] [2024-02-16 02:15:31.864885] [0x0000ffff9d65edb0] [warning]  UuidGenerator 't0000000-0000-0000-0000-00000000000g' contains invalid characters 'tg'
 [2024/02/16 13:15:32.140] [2024-02-16 02:15:31.876893] [0x0000ffff9d65edb0] [warning]  UuidGenerator '00000000-0000-0000-0000-00000000000ф' contains invalid characters 'ф'
 [2024/02/16 13:15:32.140] [2024-02-16 02:15:31.888931] [0x0000ffff9d65edb0] [warning]  UuidGenerator '0000000' is not a valid format. The format must match: '^([0-9a-f]{8}\-[0-9a-f]{4}\-[0-9a-f]{4}\-[0-9a-f]{4}\-[0-9a-f]{12})$'
 [2024/02/16 13:15:32.140] [2024-02-16 02:15:31.900994] [0x0000ffff9d65edb0] [warning]  UuidGenerator '00000000-0000-0000-0000-00000000000' is not a valid format. The format must match: '^([0-9a-f]{8}\-[0-9a-f]{4}\-[0-9a-f]{4}\-[0-9a-f]{4}\-[0-9a-f]{12})$'
 [2024/02/16 13:15:32.140] [2024-02-16 02:15:31.913151] [0x0000ffff9d65edb0] [warning]  UuidGenerator '' is not a valid format. The format must match: '^([0-9a-f]{8}\-[0-9a-f]{4}\-[0-9a-f]{4}\-[0-9a-f]{4}\-[0-9a-f]{12})$'
 [2024/02/16 13:15:32.141] 99% tests passed, 1 tests failed out of 112
 [2024/02/16 13:15:32.141] Label Time Summary:
 [2024/02/16 13:15:32.141] cast_core_test           =   0.07 sec*proc (1 test)
 ```
 and I only ask because you've previously had a lot to do with the value generator. I do tend to agree with the unit test that ф is not a valid UUID character!

jimoleary commented:
> I've restarted the currently passing base task just in case (to see if it fails), and this PR's test failed twice the same way, I currently think this is real but I'm very confused.

@thessem Those particular UUID generator tests are expected to fail see [here](https://github.com/mongodb/genny/blob/d103863204986e8caf63ef7d112d22beef7eda75/src/value_generators/test/DocumentGeneratorTestCases.yml#L1501-L1524).

AFAICS the tests that are failing are this [one](https://evergreen.mongodb.com/test_log/genny_amazon2_arm64_t_cmake_test_patch_d103863204986e8caf63ef7d112d22beef7eda75_65cec0a4c9ec4424a0fdf7c3_24_02_16_01_55_49/1?test_name=0cd66bcfb2fcebce7481eb11d2082e67&group_id=#L1) (the different values are highlighted in bold):

 [2024/02/16 02:15:32.279] FAILURE: toString(expectedFix.view()) == toString(actual.view()) (REQUIRE)
 [2024/02/16 02:15:32.279] FAILED:
 [2024/02/16 02:15:32.279]   REQUIRE( toString(expectedFix.view()) == toString(actual.view()) )
 [2024/02/16 02:15:32.279] with expansion:
 [2024/02/16 02:15:32.279]   "{ "a" : { "$numberDouble" : "3.952394090941460103" }, "uniform" : {
 [2024/02/16 02:15:32.279]   "$numberDouble" : "3.0360695791166998347" }, "exponential" : {
 [2024/02/16 02:15:32.279]   "$numberDouble" : "0.070786395988993736084" }, "gamma" : { "$numberDouble" :
 [2024/02/16 02:15:32.279]   "0.2668568736212015291" }, "weibull" : { "$numberDouble" : "0.
 [2024/02/16 02:15:32.279]   84158009772916297919" }, "extremeValue" : { "$numberDouble" : "4.
 [2024/02/16 02:15:32.279]   3175047395265959693" }, "beta" : { "$numberDouble" : "0.65807401245179697469"
 [2024/02/16 02:15:32.279]   }, "laplace" : { "$numberDouble" : "8.7629362545759938996" }, "normal" : {
 [2024/02/16 02:15:32.279]   **"$numberDouble" : "56.862224120853824161"** }, "lognormal" : { "$numberDouble"
 [2024/02/16 02:15:32.279]   : "3.3101941543832040793" }, "chisquared" : { "$numberDouble" : "8.
 [2024/02/16 02:15:32.279]   4397336386305426004" }, "cauchy" : { "$numberDouble" : "17.
 [2024/02/16 02:15:32.279]   921856487345845466" } }"
 [2024/02/16 02:15:32.279]   ==
 [2024/02/16 02:15:32.279]   "{ "a" : { "$numberDouble" : "3.952394090941460103" }, "uniform" : {
 [2024/02/16 02:15:32.279]   "$numberDouble" : "3.0360695791166998347" }, "exponential" : {
 [2024/02/16 02:15:32.279]   "$numberDouble" : "0.070786395988993736084" }, "gamma" : { "$numberDouble" :
 [2024/02/16 02:15:32.279]   "0.2668568736212015291" }, "weibull" : { "$numberDouble" : "0.
 [2024/02/16 02:15:32.279]   84158009772916297919" }, "extremeValue" : { "$numberDouble" : "4.
 [2024/02/16 02:15:32.279]   3175047395265959693" }, "beta" : { "$numberDouble" : "0.65807401245179697469"
 [2024/02/16 02:15:32.279]   }, "laplace" : { "$numberDouble" : "8.7629362545759938996" }, "normal" : {
 [2024/02/16 02:15:32.279]   **"$numberDouble" : "56.862224120853831266"** }, "lognormal" : { "$numberDouble"
 [2024/02/16 02:15:32.279]   : "3.3101941543832040793" }, "chisquared" : { "$numberDouble" : "8.
 [2024/02/16 02:15:32.279]   4397336386305426004" }, "cauchy" : { "$numberDouble" : "17.
 [2024/02/16 02:15:32.279]   921856487345845466" } }"

....

 [2024/02/16 02:15:32.279] at /data/mci/5cf9570ee54654c22db2910d4ba06a8d/src/genny/src/value_generators/test/DocGenYamlTestCaseRunner.cpp:108


Given the comment [here](https://github.com/mongodb/genny/blob/d103863204986e8caf63ef7d112d22beef7eda75/src/value_generators/test/DocumentGeneratorTestCases.yml#L825-L828), it looks like the short term fix is to disable the nornmal distribution also:

```
# The fisherf, studentt, and noncentralchi distributions give different results on different platforms.
#      fisherf: {^RandomDouble: {distribution: fisher_f, m: 12, n: 7}}
#      studentt: {^RandomDouble: {distribution: student_t, n: 17}}
#      noncentralchi: {^RandomDouble: {distribution: non_central_chi_squared, k: 7, lambda: 2}}
```

In other situations, I've seen this sort of variation dealt with by taking the difference between the actual and expected values and asserting that it's less than some limit. The actual difference of 0.000000000000007 provides a lot of room!

thessem commented:
@jimoleary thanks for this. Cursed yet again by a case of logging things that look like errors! Those other tests failing on inconsistent platforms with different optimization flags makes a lot of sense. I agree with your solve of adding a delta to the compare.

IamXander reviewed (commented):
Awesome! Any chance you have the command lines captured? Can you confirm that things are compiling with O2 or O3?

thessem commented:
@IamXander Genny is being compiled in `-O3` now.

Here's a snippet from my `compile_commands.json`:

```
{
  "directory": "/Users/james/Projects/MongoDB/genny/build",
  "command": "/Library/Developer/CommandLineTools/usr/bin/c++ -DBOOST_ALL_DYN_LINK -DCARES_STATICLIB -DYAML_CPP_STATIC_DEFINE -I<blah> -isystem /opt/data/mci/gennytoolchain/installed/arm64-osx/include -isystem <BLAH> -O3 -DNDEBUG -std=c++17 -arch arm64 -fPIC -o src/value_generators/CMakeFiles/value_generators.dir/src/DocumentGenerator.cpp.o -c <snip>/genny/src/value_generators/src/DocumentGenerator.cpp",
  "file": "<snip>/genny/src/value_generators/src/DocumentGenerator.cpp",
  "output": "src/value_generators/CMakeFiles/value_generators.dir/src/DocumentGenerator.cpp.o"
}
```

Here's a snippet from Ninja compiling in verbose mode:
```
/Library/Developer/CommandLineTools/usr/bin/c++ -DBOOST_ALL_DYN_LINK -DCARES_STATICLIB -DYAML_CPP_STATIC_DEFINE -I<snip>
 -isystem <snip> -O3 -DNDEBUG -std=c++17 -arch arm64 -isysr
oot /Library/Developer/CommandLineTools/SDKs/MacOSX14.2.sdk -fPIC -MD -MT src/canaries/CMakeFiles/canaries.dir/src/Loops.cpp.o -MF src/canaries/CMakeFiles/canaries.dir/src/Loops.cpp.
o.d -o src/canaries/CMakeFiles/canaries.dir/src/Loops.cpp.o -c /Users/james/Projects/MongoDB/genny/src/canaries/src/Loops.cpp
```

Here's a line taken out of an older build on evergreen from the CMake
```
Found OpenSSL: /data/mci/gennytoolchain/installed/arm64-linux/debug/lib/libcrypto.a (found suitable version "3.1.4", minimum required is "1.1")
```

Compared to this build which is
```
-- Found OpenSSL: /opt/data/mci/gennytoolchain/installed/arm64-osx/lib/libcrypto.a (found suitable version "3.1.4", minimum required is "1.1")
```

thessem commented:
@jimoleary and @jawwadasghar I added you as reviewers as suggested by @brad-devlugt. 

Sorry about sending this over without finished tests, it's a bit late for me right now so assume if the `cmake-tests` fail I'll take back out the flaky tests I added in (maybe this is overconfidence?).

thessem commented on src/value_generators/test/DocumentGeneratorTestCases.yml:
```diff
@@ -809,54 +809,41 @@ Tests:
-#      "noncentralchi": { "$numberDouble": "13.356763702336033717" }
+    # We'll round the results of these to 3 decimal places to make the test deterministic.
+    GivenTemplate:
+      # This first test doesn't convert to a string just to check our return type is correct
```
```suggestion
```
 
This was left in from a previous attempt

thessem commented on src/value_generators/test/DocumentGeneratorTestCases.yml:
```diff
@@ -809,54 +809,41 @@ Tests:
-#      "fisherf": { "$numberDouble": "-14.948314077879224726" }
-#      "studentt": { "$numberDouble": "-0.14294359511856821743" }
-#      "noncentralchi": { "$numberDouble": "13.356763702336033717" }
+    # We'll round the results of these to 3 decimal places to make the test deterministic.
```
```suggestion
    # R the results to 3 decimal places to make the test deterministic across platforms and compiler settings
```

thessem commented on src/value_generators/test/DocumentGeneratorTestCases.yml:
```diff
@@ -809,54 +809,40 @@ Tests:
-#      "fisherf": { "$numberDouble": "-14.948314077879224726" }
-#      "studentt": { "$numberDouble": "-0.14294359511856821743" }
-#      "noncentralchi": { "$numberDouble": "13.356763702336033717" }
+    # R the results to 3 decimal places to make the test deterministic across platforms and compiler settings
```
```suggestion
    # Round the results to 3 decimal places to make the test deterministic across platforms and compiler settings
```

jimoleary reviewed (approved):
LGTM on the changes ... but do we have any idea of the scale of changes points that this will introduce.

As this change will affect all versions that use genny, should we defer committing this change until v7.3.0 is released?

IamXander commented:
I would really push us to release this ASAP. I agree with you that we want to protect 7.3 but I think the much bigger fish is 8.0 and getting real correct data for 8.0 ASAP is worth it. Open to having my mind changed though

thessem commented:
We don't version genny (yet) so unfortunately this change is for all branches of sys-perf.

I was going to give @jawwadasghar a chance to weigh in on that aspect. I only ran my patches with the high value workloads and it looks like quite a few of the genny ones have fairly significant changes. I don't have any guesses about how I've changed the noise profiles, I figured I'd have to run many more test repetitions to get an idea of that and it didn't seem like useful (as in it would go or no-go this PR) information anyway.

I was going to hit merge when I had two ticks but it seems like the consensus is to wait until 7.3.0 is released?

thessem commented on src/value_generators/test/DocumentGeneratorTestCases.yml:
```diff
@@ -809,54 +809,40 @@ Tests:
-#      "fisherf": { "$numberDouble": "-14.948314077879224726" }
-#      "studentt": { "$numberDouble": "-0.14294359511856821743" }
-#      "noncentralchi": { "$numberDouble": "13.356763702336033717" }
+    # Round the results to 3 decimal places to make the test deterministic across platforms and compiler settings
```
Thanks :) I think we should actually add a `round` (or at least, number of bytes in final output, given how weird floats get get) and a `seed` parameter to these random generators at some point. Having a `seed` makes them easier to test (and replicate?) and a round might be handy for the end-user too.

jimoleary commented:
I was chatting with @bsnyman about the upcoming 7.3.0 release / code freeze:

1. The code freeze for 7.3.0 is expected to last until March 26th 
2. The 7.3.0 GA binaries are expected to be built March 5th

It would likely be fairly disruptive to change many genny results in the middle of the release cycle and pinning the genny version in 7.3 runs the risk of not being able to pick up a necessary genny fix (although I'm not sure this is a big risk).

Is March 5th a reasonable compromise to delay this merge?

jawwadasghar commented:
Change LGTM. 
Looking at perf analyzer, I only see a handful of changes when filtering for z-score > 2, which I believe is our cutoff. 
They all seem like an improvements (decreasing latency or increasing throughput), so seems like those workloads were bound by the workload client, which seems bad. 
Jim's idea of submitting March 5th seems like a fine compromise to me. 
![image](https://github.com/mongodb/genny/assets/112432173/38dc69d9-4be2-4c9a-9282-260f3e7b0b32)

brad-devlugt commented:
I think if we move forward with this we'll have more accurate test results for 7.3. What work will this add to the 7.3 release cycle? Will we need to do a pass through the perf-analyzer to validate the changes are as expected based @thessem's testing or is there additional work I'm missing here?

thessem commented:
I filtered on .5 because at .75 I'm starting to see that _maybe_ we had an effect
![image](https://github.com/mongodb/genny/assets/973593/02ea885f-8e88-4133-8429-aa2a83070493)

It is also hard to say for sure (I only ran 4 tests) but it seems like there are some noise reductions at around the .6 stddevs too, e.g.:
![image](https://github.com/mongodb/genny/assets/973593/e878f8bf-9d3a-41b1-b1d7-116c3eb4e45c)
![image](https://github.com/mongodb/genny/assets/973593/63fc5db3-a189-488f-ac44-138b27fc6476)

I think we need to make a choice pretty soon about merging this in today, or waiting until 7.3 is released. I don't have the context to make that decision.

IamXander commented:
These changes only effect the genny libraries right? I think that likely this might not have a huge effect which is why you are just seeing noise. The most important thing to know before merging this is are we creating additional noise? If not we should move forward with this.

thessem commented:
evergreen merge
